src/pox-helpers.ts

```diff
@@ -221,31 +247,360 @@ function networkToHrp(network: 'mainnet' | 'testnet' | 'regtest'): string {
+}>;
+
+type PoX2EventData = ClarityValueTuple<{
+  name: ClarityValueStringAscii;
+  balance: ClarityValueUInt;
+  stacker: ClarityValuePrincipalStandard | ClarityValuePrincipalContract;
+  locked: ClarityValueUInt;
+  'burnchain-unlock-height': ClarityValueUInt;
+  data: ClarityValueTuple;
+}>;
+
+export const enum Pox2EventName {
+  HandleUnlock = 'handle-unlock',
+  StackStx = 'stack-stx',
+  StackIncrease = 'stack-increase',
+  StackExtend = 'stack-extend',
+  DelegateStackStx = 'delegate-stack-stx',
+  DelegateStackIncrease = 'delegate-stack-increase',
+  DelegateStackExtend = 'delegate-stack-extend',
+  StackAggregationCommit = 'stack-aggregation-commit',
+}
+
+interface Pox2PrintEventTypes {
+  [Pox2EventName.HandleUnlock]: {
+    'first-cycle-locked': ClarityValueUInt;
+    'first-unlocked-cycle': ClarityValueUInt;
+  };
+  [Pox2EventName.StackStx]: {
+    'lock-amount': ClarityValueUInt;
+    'lock-period': ClarityValueUInt;
+    'pox-addr': Pox2Addr;
+    'start-burn-height': ClarityValueUInt;
+    'unlock-burn-height': ClarityValueUInt;
+  };
+  [Pox2EventName.StackIncrease]: {
+    'increase-by': ClarityValueUInt;
+    'total-locked': ClarityValueUInt;
+  };
+  [Pox2EventName.StackExtend]: {
+    'extend-count': ClarityValueUInt;
+    'unlock-burn-height': ClarityValueUInt;
+    'pox-addr': Pox2Addr;
+  };
+  [Pox2EventName.DelegateStackStx]: {
+    'lock-amount': ClarityValueUInt;
+    'unlock-burn-height': ClarityValueUInt;
+    'pox-addr': Pox2Addr;
+    'start-burn-height': ClarityValueUInt;
+    'lock-period': ClarityValueUInt;
+    delegator: ClarityValuePrincipalStandard | ClarityValuePrincipalContract;
+  };
+  [Pox2EventName.DelegateStackIncrease]: {
+    'pox-addr': Pox2Addr;
+    'increase-by': ClarityValueUInt;
+    'total-locked': ClarityValueUInt;
+    delegator: ClarityValuePrincipalStandard | ClarityValuePrincipalContract;
+  };
+  [Pox2EventName.DelegateStackExtend]: {
+    'pox-addr': Pox2Addr;
+    'unlock-burn-height': ClarityValueUInt;
+    'extend-count': ClarityValueUInt;
+    delegator: ClarityValuePrincipalStandard | ClarityValuePrincipalContract;
+  };
+  [Pox2EventName.StackAggregationCommit]: {
+    'pox-addr': Pox2Addr;
+    'reward-cycle': ClarityValueUInt;
+    'amount-ustx': ClarityValueUInt;
+  };
+}
+
+function clarityPrincipalToFullAddress(
+  principal: ClarityValuePrincipalStandard | ClarityValuePrincipalContract
+): string {
+  if (principal.type_id === ClarityTypeID.PrincipalStandard) {
+    return principal.address;
+  } else if (principal.type_id === ClarityTypeID.PrincipalContract) {
+    return `${principal.address}.${principal.contract_name}`;
+  }
+  throw new Error(
+    `Unexpected Clarity value type for principal: ${(principal as ClarityValue).type_id}`
+  );
+}
+
+// TODO: this and the logic referencing it can be removed once the "stale" data issue is fixed in
+// https://github.com/stacks-network/stacks-blockchain/pull/3318
+const PATCH_EVENT_BALANCES = true;
```

Comment:
Note for reviewers:
The pox2 events emitted are currently bugged with an issue where the locked state (balance, locked amount, unlock height) are all "stale" and report the state from _before_ the transaction. The code checking this value is performing a hacky workaround where operation-specific args are used to "patch" the stale values into the correct state.

Reply:
It was decided to keep this behavior, at least for now.